src/bundle/Security/Http/Authenticator/TwoFactorAuthenticator.php

```diff
@@ -0,0 +1,228 @@
+        'check_path' => TwoFactorFactory::DEFAULT_CHECK_PATH,
+        'auth_code_parameter_name' => TwoFactorFactory::DEFAULT_AUTH_CODE_PARAMETER_NAME,
+        'trusted_parameter_name' => TwoFactorFactory::DEFAULT_TRUSTED_PARAMETER_NAME,
+    ];
+
+    /**
+     * @var TokenStorageInterface
+     */
+    private $tokenStorage;
+
+    /**
+     * @var HttpUtils
+     */
+    private $httpUtils;
+
+    /**
+     * @var string
+     */
+    private $firewallName;
+
+    /**
+     * @var AuthenticationSuccessHandlerInterface
+     */
+    private $successHandler;
+
+    /**
+     * @var AuthenticationFailureHandlerInterface
+     */
+    private $failureHandler;
+
+    /**
+     * @var AuthenticationRequiredHandlerInterface
+     */
+    private $authenticationRequiredHandler;
+
+    /**
+     * @var string[]
+     */
+    private $options;
+
+    /**
+     * @var EventDispatcherInterface
+     */
+    private $eventDispatcher;
+
+    /**
+     * @var LoggerInterface
+     */
+    private $logger;
+    /**
+     * @var TwoFactorAccessDecider
+     */
+    private $twoFactorAccessDecider;
+
+    public function __construct(
+        TokenStorageInterface $tokenStorage,
+        HttpUtils $httpUtils,
+        string $firewallName,
+        AuthenticationSuccessHandlerInterface $successHandler,
+        AuthenticationFailureHandlerInterface $failureHandler,
+        AuthenticationRequiredHandlerInterface $authenticationRequiredHandler,
+        array $options,
+        TwoFactorAccessDecider $twoFactorAccessDecider,
+        EventDispatcherInterface $eventDispatcher,
+        ?LoggerInterface $logger = null
+    ) {
+        if (empty($firewallName)) {
+            throw new \InvalidArgumentException('$firewallName must not be empty.');
+        }
+
+        $this->tokenStorage = $tokenStorage;
+        $this->httpUtils = $httpUtils;
+        $this->firewallName = $firewallName;
+        $this->successHandler = $successHandler;
+        $this->failureHandler = $failureHandler;
+        $this->authenticationRequiredHandler = $authenticationRequiredHandler;
+        $this->options = array_merge(self::DEFAULT_OPTIONS, $options);
+        $this->twoFactorAccessDecider = $twoFactorAccessDecider;
+        $this->eventDispatcher = $eventDispatcher;
+        $this->logger = $logger ?? new NullLogger();
+    }
+
+    public function supports(Request $request): ?bool
+    {
+        if (!$this->isCheckAuthCodeRequest($request)) {
+            return false;
+        }
+
+        $currentToken = $this->tokenStorage->getToken();
+//        return $this->isCheckAuthCodeRequest($request)
+//            || $this->isAuthFormRequest($request)
+//            // Let routes pass, e.g. if a route needs to be callable during two-factor authentication
+//            || !$this->twoFactorAccessDecider->isAccessible($request, $currentToken);
+
+        return $currentToken instanceof TwoFactorTokenInterface && $currentToken->getProviderKey() === $this->firewallName;
```

Comment:
Is this `$currentToken->getProviderKey() === $this->firewallName` necessary or does the new system guarantee that only authenticators from the firewall itself are called?

Reply:
Yes, it does (the new system has one `AuthenticatorManager` instance per firewall, so it's only aware of the authenticators for that specific firewall)

Reply:
Great, another parameter that I can remove from the constructor :)